File: src/main/java/com/nugurang/graphql/resolver/NotificationResolver.java
package com.nugurang.graphql.resolver;

import com.nugurang.dao.NotificationDao;
import com.nugurang.dao.NotificationDataDao;
import com.nugurang.dto.NotificationDto;
import com.nugurang.dto.NotificationTypeDto;
import com.nugurang.dto.UserDto;
import graphql.kickstart.tools.GraphQLResolver;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

@Service
public class NotificationResolver implements GraphQLResolver<NotificationDto> {
    private final NotificationDao notificationDao;
    private final NotificationDataDao notificationDataDao;

    public UserDto user(NotificationDto notificationDto) {
        return notificationDao.findById(notificationDto.getId()).get().getUser().toDto();
    }

    public List<String> data(NotificationDto notificationDto) {
        return notificationDataDao.findAllByNotificationIdOrderByIdAsc(notificationDto.getId()).stream().map(entity -> entity.getValue()).collect(Collectors.toList());
    }

    public NotificationTypeDto type(NotificationDto notificationDto) {
        return notificationDao.findById(notificationDto.getId()).get().getNotificationType().toDto();
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public NotificationResolver(final NotificationDao notificationDao, final NotificationDataDao notificationDataDao) {
        this.notificationDao = notificationDao;
        this.notificationDataDao = notificationDataDao;
    }
    //</editor-fold>
}

File: src/main/kotlin/com/nugurang/graphql/resolver/NotificationResolver.kt
package com.nugurang.graphql.resolver

import com.nugurang.dao.NotificationDao
import com.nugurang.dao.NotificationDataDao
import com.nugurang.dto.NotificationDto
import com.nugurang.dto.NotificationTypeDto
import com.nugurang.dto.UserDto
import com.nugurang.entity.NotificationTypeEntity
import com.nugurang.entity.UserEntity
import com.nugurang.exception.NotFoundException
import com.nugurang.mapper.NotificationTypeMapper
import com.nugurang.mapper.UserMapper
import graphql.kickstart.tools.GraphQLResolver
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class NotificationResolver(
    private val notificationDao: NotificationDao,
    private val notificationDataDao: NotificationDataDao,
    private val notificationTypeMapper: NotificationTypeMapper,
    private val userMapper: UserMapper
) : GraphQLResolver<NotificationDto> {

    fun user(notificationDto: NotificationDto): UserDto {
        return notificationDao.findByIdOrNull(notificationDto.id)?.user?.let(userMapper::toDto)
        ?: throw NotFoundException(UserEntity::class.java)
    }

    fun data(notificationDto: NotificationDto): List<String> {
        return notificationDataDao.findAllByNotificationIdOrderByIdAsc(notificationDto.id).map { it.value }
    }

    fun type(notificationDto: NotificationDto): NotificationTypeDto {
        return notificationDao.findByIdOrNull(notificationDto.id)?.notificationType?.let(notificationTypeMapper::toDto)
        ?: throw NotFoundException(NotificationTypeEntity::class.java)
    }
}